Free Trials: Move free trial checkout to a new controller

Tug opened the pull request:
Move the free trial checkout to its own controller. This has the advantage of lightening the checkout component complexity and fixing a few minor bugs which made the free trial upgrade conflict with other cart items.

#### Context
See this comment https://github.com/Automattic/wp-calypso/issues/286#issuecomment-171041340

Should fix #1538 and #2207 and #2274 as well.

### Testing instructions
`git checkout update/free-trial-checkout-page`

#### Test 1
- Navigate to http://calypso.localhost:3000/plans/:site for a site eligible to free trial
- Click on `Start Free Trials` and check that you are redirected to the Free Trial confirmation page.
- Confirm subscription and check that you are redirected to the Plans page and that the thank you notice is displayed.
- Close the notice

#### Test 2 (needs a fix in wpcom)
- Add a product (such as a premium theme) to your cart.
- Don't complete checkout yet but go to the plans page and start a free trial.
- Complete checkout of your product.

#### Test 3
- Navigate to `/start/free-trial`, choose a theme and a name for your new site.
- On the Plans page click on one of the `Start Free Trials` button.
- Ensure you are redirected to `/plans/:site/thank-you` and that the thank you notice is displaying.

#### Test 4 (logged out)
- Navigate to `/start/free-trial`
- Enter `localStorage.setItem( 'ABTests', '{"freeTrials_20160112":"offered"}' );` in the console to add yourself to the test
- Choose a theme and a name for your new site.
- On the Plans page click on one of the `Start Free Trials` button.
- Choose a name and password for your new account.
- Confirm your free trial subscription
- Ensure that you are redirected to `/plans/:site/thank-you` and that the thank you notice is displaying (along with the email confirmation notice).

#### Test 5
- Cancel a free trial from the admin tools (Only available for automatticians)
- Go to the Plans page and check that the other plan is still available for a free trial.
- Click on `Start Free Trials`
- Change the url to the other plan and reload the page (for instance `/start-trial/premium/:site` becomes `/start-trial/business/:site`
- Assert that you have a message telling you you cannot purchase and inviting you to go back to the Plans page

#### Reviews
- [ ] Code Review
- [ ] Design Review
- [ ] Product Review

gziolo commented on client/my-sites/trials/controller.jsx:
```diff
@@ -0,0 +1,11 @@
+import ReactDom from 'react-dom';
+import React from 'react';
+
+module.exports = {
```
We should rather use `export {` here to keep ES6 style.

stephanethomas commented on client/sections.js:
```diff
@@ -61,6 +61,11 @@ sections = [
 	},
 	{
+		name: 'start-trial',
+		paths: [ '/start-trial' ],
```
I like the fact that we don't touch the `<Checkout />` component nor the upgrades controller whose refactoring would be time-consuming but I feel that `/checkout/trial` would be more appropriate as url. What were the reasons for opting for `start-trial`?

scruffian commented on client/my-sites/upgrades/index.js:
```diff
@@ -289,5 +290,14 @@ module.exports = function() {
 		);
+
+		page(
+			'/start-trial/:domain/:plan_name?',
```
We tend to favour having the domain at the end of the URL unless there is a compelling reason not to...

scruffian replied:
I also wonder if we need to get this URL whitelisted to point to Calypso, in case someone wants to link to it directly from an external source.

Tug replied:
Then `plan_name` won't be optional but I'm find with this.
So `start-trial/:plan_name/:domain` ?
Whitelisted, you mean removing the `isEligible` middleware?

scruffian replied:
> Then plan_name won't be optional but I'm find with this.

It can't be really, as we need to know what plan they want

>So /start-trial/:plan_name/:domain ?

:+1: 

> Whitelisted, you mean removing the isEligible middleware?

No. This is something that we need to do on the server, if we want to.

stephanethomas replied:
I think the convention is to use `:site` over `:domain` for routes.

Moreover I'm pretty sure we should camel case parameters (i.e. `:planName` instead of `:plan_name`).

scruffian commented on client/my-sites/upgrades/trials/start-trial/free-trial-confirmation-box.js:
```diff
@@ -35,6 +35,15 @@ const FreeTrialConfirmationBox = React.createClass( {
 				<div className="payment-box-actions">
 					<PayButton
+						beforeSubmitText={
+							this.translate( 'Start %(days)s Day Free Trial', {
```
Does this need to pass 14 in the args like this? I think the reason for doing that is if this number will change programatically, which this one won't...

Tug replied:
I have just moved this part from `checkout/pay-button.jsx` so I didn't take any decision on this. But I personally think it makes the translation more generic. Any reason not to use arguments?

scruffian replied:
It adds complexity for translators. I don't feel strongly.

scruffian commented on client/my-sites/upgrades/trials/start-trial/terms-of-service.jsx:
```diff
@@ -0,0 +1,31 @@
+/**
```
Do we need a new component here? Can't we reuse `client/my-sites/upgrades/checkout/terms-of-service.jsx`?

Tug replied:
I have just changed `By checking out` to `By continuing` since we were trying to avoid showing anything related to a purchase. We could definitively use `checkout/terms-of-service.jsx` to avoid code duplication if you think it's not worth it.

scruffian replied:
I think we should use it and just pass a prop to change the copy.

Tug replied:
There is also this analytics event which differs

scruffian replied:
We could also pass that as a prop...

fabianapsimoes commented:
`/start/free-trial/plans` is broken in this branch -- it won't load past the placeholders. Here is the error I see in the console:

`TypeError: this.props.site is undefined`

For the record `/start/plans` still works fine.

scruffian commented on client/components/data/transaction/README.md:
```diff
@@ -0,0 +1,28 @@
+CartData
```
Should this say `TransactionData`?

Tug replied:
Oops, I'm not using this module anymore but created a module to handle both a temporary cart and the transaction store here `client/my-sites/upgrades/trials/data/free-trial-cart.jsx`.
So I'm going to remove `client/components/data/transaction`.

scruffian commented on client/my-sites/upgrades/checkout/pay-button.jsx:
```diff
@@ -113,15 +105,16 @@ var PayButton = React.createClass( {
-		} else {
-			text = this.translate( 'Completing your purchase', { context: 'Loading state on /checkout' } )
+		if ( this.props.completingText ) {
+			return {
```
I guess this tab is a mistake?

Tug replied:
Weird, I do have a space on my end. I think the github diff tool has some limitations. If you look here it does look fine https://github.com/Automattic/wp-calypso/blob/update/free-trial-checkout-page/client/my-sites/upgrades/checkout/pay-button.jsx#L109

fabianapsimoes commented:
After clicking the confirmation button in `/start-trial`, there's a brief glimpse of the regular version of `/plans`, before it converts to the Trial Hub. Not a big deal, but it'd be nice to skip that.

I tested the scenario in #1538, where you complete a Free Trial checkout when you have another upgrade in the cart. It works fine, but the cart is empty after you confirm the start of the trial, so it doesn't solve the original problem mentioned in #286.

This solves #1538 and #2207, as promised.

scruffian commented on client/my-sites/upgrades/trials/data/free-trial-cart.jsx:
```diff
@@ -0,0 +1,91 @@
+	mixins: [ observe( 'sites', 'products' ) ],
+
+	getInitialState: function() {
+		return { cart: {} };
```
Is calling this "cart" a good idea, since this isn't a cart in the same way that our other "cart"s are (unless I am misunderstanding)

Tug replied:
It is a cart. It has the same structure, can be given to any `cart-values` function... It's just that it doesn't live in the `CartStore` so it really is temporary. The idea is to have this component handle the conversion from a `planName` (`premium` or `business`) to a valid cart object which will be given as props to its children and used in `checkout/transaction-steps-mixin.jsx` to make the actual checkout request. And since it also needs a valid `transaction` props this component also handles that.

scruffian replied:
Makes sense :+1:

scruffian commented on client/my-sites/upgrades/trials/start-trial/terms-of-service.jsx:
```diff
@@ -0,0 +1,31 @@
+	displayName: 'TermsOfService',
+
+	recordTermsAndConditionsClick: function() {
+		analytics.ga.recordEvent( 'Tree Trials', 'Clicked Terms and Conditions Link' );
```
What are `Tree Trials`? ;)

Tug replied:
Typo in `Tree Trials`

Tug commented:
@fabianapsimoes Thanks for testing all this!
Let's add your test of the signup in the test cases and I'll fix it for the next review.
The invalidation on `sitePlans` wasn't added here yet, but it's planned for this PR as well.

Tug commented:
I have addressed most of the issues you raised here. I think it's ready for a second review.

fabianapsimoes commented:
I'm getting a different error on `/start-trial` when coming from `/start/free-trial` as a new user:

![screenshot-calypso localhost 3000 2016-01-18 17-58-28](https://cloud.githubusercontent.com/assets/1248436/12397489/c499d5be-be0d-11e5-89f4-9653cd6567da.png)

> I tested the scenario in #1538, where you complete a Free Trial checkout when you have another upgrade in the cart. It works fine, but the cart is empty after you confirm the start of the trial, so it doesn't solve the original problem mentioned in #286.

This is still an issue as well. But I wouldn't block on it to merge this PR.

---
Tested cases:

- `/start` choosing free plan
- `/start` choosing paid plan
- `/start/free-trial` choosing free plan (new user)
- `/start/free-trial` choosing paid plan (new user)
- `/start/free-trial` choosing free plan (existing user)
- `/start/free-trial` choosing paid plan (existing user)
- trial checkout when you have another product in the cart

fabianapsimoes commented:
Also noting we might have to wait until the test is over to merge this. Otherwise, we might affect the results of the test since this change hides the cart in checkout. See p1453141195000050-slack-upgrades.

Tug commented:
Thanks for testing once again @fabianapsimoes 
I have fixed the bug for new users and added this as a 4th test case in the instructions.
I was not able to fix the issue with the temporary cart being emptied on checkout and I believe this require a fix to the WP-API.

fabianapsimoes commented:
- I've noticed a regression in this PR. It's again taking a while for the sidebar label to update from Free to Trial.

- I'm seeing this error in the console when I go through `/start/free-trial/plans` for an existing user:
`TypeError: this.props.sitePlans is undefined`

- I've noticed a weird error in logged-in signup via `/start/free-trial`. Steps to reproduce:
 - With an existing account, create a new site. You'll see the error I mentioned above, but the site will be create successfully.
 - Try creating a new site again. You might see that error once more, and will be able to move through the flow until the processing screen. However, the processing screen will be stuck in "Setting up your site".

>  I was not able to fix the issue with the temporary cart being emptied on checkout and I believe this require a fix to the WP-API.

We should look into this in a separate issue then. I don't think it's a very important case.

Tug commented:
@fabianapsimoes Thanks for testing again.
> I've noticed a regression in this PR. It's again taking a while for the sidebar label to update from Free to Trial.
I don't think this was ever fixed. We just trigger a new fetch of `/sites` so the time it takes depends on the response time from the server.

I'm going to have a look at the second error you mentioned.

stephanethomas commented on client/my-sites/plans/controller.jsx:
```diff
@@ -17,15 +17,22 @@ var sites = require( 'lib/sites-list' )(),
+		}
+
+		upgradesActions.addItem( cartItem );
+		page( `/checkout/${ siteSlug }` );
```
We could probably use an `else` clause here:

```javascript
if ( productValues.isFreeTrial( cartItem ) ) {
  page( `/start-trial/${ plans.getPathFromSlug( cartItem.product_slug ) }/${ siteSlug }` );
} else {
  upgradesActions.addItem( cartItem );
  
  page( `/checkout/${ siteSlug }` );
}
```

That way we get rid of the `return` statement and the whole block is more readable.

Tug replied:
I feel both syntax are equally readable. The advantage I see to the `if () { return; }` is to avoid too many nested blocks as we add more code (and have cleaner diffs). Other than that I can use `if else` if you really prefer.

scruffian replied:
I prefer @Tug's way :)

My reasoning is that when you are reading the code, once you hit the return, you know nothing else will happen. I believe our linter also enforces no unnecessary else statements, which means this should be a common pattern elsewhere in the code.

stephanethomas commented on client/my-sites/upgrades/index.js:
```diff
@@ -284,10 +285,19 @@ module.exports = function() {
 		);
+
+		page(
+			'/start-trial/:planName/:domain',
```
I feel I already mentioned it but I'm pretty sure we should use `:site` instead of `:domain` in our routes.

Tug replied:
I don't feel strongly about `:domain` but it is used everywhere for paths in this file https://github.com/Automattic/wp-calypso/blob/update/free-trial-checkout-page/client/my-sites/upgrades/index.js
So I kept it for consistency.

stephanethomas replied:
Have another look to the file you just linked and you'll see why I proposed this change. It contains routes for the domain management section and most of them use `:site` **and** `:domain` because we have to make the distinction between the site selected by the user and the domain being managed. Further down this file we have routes for the domain search section and this is where we got it wrong (it was one of the first section of the store migrated to Calypso). We should fix these but in the meantime let's try to use the right parameter for new routes :).

Tug replied:
+1, makes sense to use `:site` for site and `:domain` for domain. Would be nice to fix it later as you said

stephanethomas replied:
The reason why this is highly misleading is that you can use a domain for `:site`, i.e. when a user use a custom domain as primary address. No surprise why we managed to screw that part up :).

stephanethomas commented on client/my-sites/upgrades/trials/controller.jsx:
```diff
@@ -0,0 +1,83 @@
+import productsList from 'lib/products-list';
+import { getABTestVariation } from 'lib/abtest';
+import Main from 'components/main';
+import { setSection } from 'state/ui/actions';
```
We should order all these imports alphabetically.

Tug replied:
Not easy to sort by paths of group of submodules. I decided to sort by paths and updated the PR!

stephanethomas replied:
The idea is really to sort by **import** and not by path:

```javascript
import { getABTestVariation } from 'lib/abtest';
import i18n from 'lib/mixins/i18n';
import Main from 'components/main';
import plansList from 'lib/plans-list';
import productsList from 'lib/products-list';
import route from 'lib/route';
import { setSection } from 'state/ui/actions';
import sitesList from 'lib/sites-list';
import titleActions from 'lib/screen-title/actions';
```

I agree it's not always straightforward when using named imports though.

stephanethomas commented on client/my-sites/upgrades/trials/data/free-trial-cart.jsx:
```diff
@@ -0,0 +1,93 @@
+		this.setState( { cart: newCart } );
+	},
+
+	isLoading: function() {
```
We've been naming these methods `isDataLoading()` in other components so maybe we should follow this pattern here.

Tug replied:
Thanks for pointing this out, this has been addressed ;)

stephanethomas commented on client/my-sites/upgrades/trials/start-trial/free-trial-form.jsx:
```diff
@@ -0,0 +1,86 @@
+	},
+
+	isLoading: function() {
+		return ! cartItems.hasFreeTrial( this.props.cart );
```
That's an unexpected way of checking if data is available. Maybe we should add a comment explaining why we followed this approach here?

Tug replied:
Since this component mimics the `secure-payment-form` from checkout which handles a `cart` and a `transaction` and will receive updates on `cart` change, it seems logical that the component is ready once its `cart` contains the expected item. I'm going to add a comment to make it clear ;)

fabianapsimoes commented:
I'm unable to start a trial with a new account. Here's what happens:

- Coming from `/start/free-trial` while logged out, I'm going from the processing screen to `/plans/:site` without seeing the confirmation in `/start-trial`.
- I do see `/start-trial` in my address bar for a bit, but I never see the content of that page.
- I can see a flash of the Site Selection screen (`/plans`) before landing in `/plans/:site`.
- The trial is not added to the site I created, so `/plans/:site` is just the regular Plans page instead of the Trial Hub.

![domain](https://cloud.githubusercontent.com/assets/1248436/12446349/cd4fc968-bf69-11e5-842b-b1a2b7d19a76.gif)

Signup for logged-in users works fine.

---

Minor issues:

In case 5, I see the regular confirmation box before it changes to the message that lets me know the plan isn't available. Clicking on "Start Trial" before it updates to "Return to the Plans page" will display a red notice saying "Insufficient credits". Would it be possible to display the correct message right away? If that would be too much work, I wouldn't block this PR on it since this is a bit of an edge case. We can open a new issue and address this at a later point then.

We should also update the messaging in case 5. Right now, it says "The premium plan is not available for your site". We need to capitalize the plan name, and mention it's a trial. Maybe: "A Premium plan free trial isn't available for your site"? Paging @mikeshelton1503 for a sanity check.

Tug commented:
Wow that was tricky bug! I had to fix the `siteSelection` middleware since it can emit a `'change'` event without the sites actually being loaded. Please review the code on this since it touches a core component of calypso.

About the first minor issue, the confirmation box blink, I was aware of it but decided to leave it that way because `sitePlans` is not needed for the actual checkout (it's just needed to check that you can actually get a free trial without error). We can either:
- Leave it that way and only users who are not eligible and land on this path by mistake (so very few users) will see the blink
- Show the placeholder while `sitePlans` loads and it may take 2 or 3 more seconds to display.

About the second minor issue, I'm going to make a quick fix for this as you suggested and we can discuss further improvements here.

Tug commented:
This is how it looks like now:
![image](https://cloud.githubusercontent.com/assets/230230/12450700/8c57764c-bf85-11e5-84a5-585ce6053361.png)

fabianapsimoes commented:
I'm getting stuck in `/start-trial` after the processing screen. Again, this happens for new users only. Not sure if this happens every time a new user goes through that flow (I'm under the impression I managed to complete it a few times, but maybe I did something different), but it definitely happens a lot.

![domain](https://cloud.githubusercontent.com/assets/1248436/12453540/f8fb5512-bf93-11e5-9ed2-c84aac522a26.gif)

Tug commented on client/components/plans/plan-list/index.jsx:
```diff
@@ -46,12 +46,12 @@ module.exports = React.createClass( {
-		}
-
 		if ( ! this.props.isInSignup ) {
+			if ( this.props.sites ) {
```
This change improves clarity in children components because we don't want `site` to be available for them during signup.

Tug commented on client/my-sites/controller.js:
```diff
@@ -147,15 +148,19 @@ module.exports = {
+			onSitesChange = function() {
+				// if sites have not loaded yet, wait for next change
+				if ( ! sites.fetched ) return;
+				sites.off( 'change', onSitesChange );
```
It should improve things because `sites` can emit multiple `'change'` events after data has been fetched from the server.

mikeshelton1503 commented:
For the Case 5 copy, the edits @Tug made look pretty good. I only have a couple of adjustments to suggest:

> The WordPress.com Premium free trial is not available
> Sorry, you have already tried the WordPress.com Premium free trial for [sitename] however you may purchase the WordPress.com Premium plan if you'd like.

Then for the buttons let's add a "Purchase WordPress.com Premium" primary button and change the "Return to Plans page" button to secondary with this copy "View All Plans"

Tug commented:
Ok I was mistaken by the `siteSelection` middleware (it does not always reload the page) so I removed that commit from the PR and then simply moved the checks for a current free trial out of the `isEligible` middleware. It shouldn't cause any loading time issue now.
I propose we review this PR again and decide how it should be merged.
If we go later for the a checkout with a free trial and a domain name at the same time, maybe we need to decouple the UI from the actual transaction request (the 2 are really tied together at the moment with the whole `TransactionStepsMixin` thing).

fabianapsimoes commented:
We're currently discussing a few changes to how Free Trials work (p6qnuF-1ly-p2) that might be conflicting to this PR, so I'm setting it to Punted until we know more. These changes revolve around the idea of offering both trials and domains in `/start`, which might require the cart / regular checkout flow to be used.

stephanethomas commented on client/state/sites/plans/actions.js:
```diff
@@ -25,12 +25,10 @@ import {
  * @returns {Function} the corresponding action thunk
```
The docblock comment should be updated.

fabianapsimoes commented:
@lancewillett we'll get back to this PR soon, as we start working on Free Trials again. Reopening.
